### PR TITLE
Add support for mounted GNOME Trash folders

### LINKED ISSUE #908: Search mounts for GNOME Trash folders
#852 introduced the GNOME Trash plugin. It currently only searches for Trash folders inside user homes, but this application also leaves traces on mounted devices.

### PR DESCRIPTION
**Pull request by JSCU-CNI**
This PR adds mounted Trash folder support to the GNOME Trash plugin. Fixes #908.

### CODE REVIEW

File: dissect/target/filesystem.py [hunk 1]
```diff
@@ -1300,6 +1300,8 @@ def symlink(self, src: str, dst: str) -> None:
 class LayerFilesystem(Filesystem):
     __type__ = "layer"
 
+    mounts: dict = {}
```
Comment (Schamper) [inline dissect/target/filesystem.py:1303]: Is this addition only for the unit test? I'd prefer finding a more suitable way to mock in that case :smile:.

**Review by Schamper — CHANGES_REQUESTED** [review]

File: dissect/target/plugins/os/unix/trash.py [hunk 2]
```diff
@@ -31,15 +31,25 @@ class GnomeTrashPlugin(Plugin):
...
         """it aint much, but its honest work"""
+
+        # home trash folders
         for user_details in self.target.user_details.all_with_home():
             for trash_path in self.PATHS:
                 if (path := user_details.home_path.joinpath(trash_path)).exists():
                     yield user_details, path
 
+        # mounted devices trash folders
+        for mount_path in list(self.target.fs.mounts.keys()) + ["/mnt", "/media"]:
```
Comment (Horofic) [inline dissect/target/plugins/os/unix/trash.py:46]:
```suggestion
        for mount_path in list(self.target.fs.mounts) + ["/mnt", "/media"]:
```

File: dissect/target/plugins/os/unix/trash.py [hunk 3]
```diff
@@ -31,15 +31,25 @@ class GnomeTrashPlugin(Plugin):
...
+            if mount_path == "/":
+                continue
+
+            for mount_trash in self.target.fs.path(mount_path).glob("**/.Trash-*"):
+                yield None, mount_trash
```
Comment (Horofic) [inline dissect/target/plugins/os/unix/trash.py:51]:
```suggestion
                yield UserDetails(None, None), mount_trash
```

File: dissect/target/plugins/os/unix/trash.py [hunk 4]
```diff
@@ -96,7 +107,7 @@ def trash(self) -> Iterator[TrashRecord]:
                             filesize=file.lstat().st_size if file.is_file() else None,
                             deleted_path=file,
                             source=trash_info_file,
-                            _user=user_details.user,
+                            _user=user_details.user if user_details else None,
```
Comment (Horofic) [inline dissect/target/plugins/os/unix/trash.py:110]:
```suggestion
                            _user=user_details.user
```

File: dissect/target/plugins/os/unix/trash.py [hunk 5]
```diff
@@ -110,7 +121,7 @@ def trash(self) -> Iterator[TrashRecord]:
                         filesize=0,
                         deleted_path=deleted_path,
                         source=trash_info_file,
-                        _user=user_details.user,
+                        _user=user_details.user if user_details else None,
```
Comment (Horofic) [inline dissect/target/plugins/os/unix/trash.py:124]:
```suggestion
                        _user=user_details.user

**Review by Horofic — CHANGES_REQUESTED** [review]

File: tests/plugins/os/unix/test_trash.py [hunk 6]
```diff
@@ -76,3 +77,29 @@ def test_gnome_trash(target_unix_users: Target, fs_unix: VirtualFilesystem) -> N
     assert results[1].path is None
     assert results[1].deleted_path == "/home/user/.local/share/Trash/expunged/123456789/some-dir/some-file.txt"
     assert results[1].filesize == 79
+
+
+@patch(
+    "dissect.target.filesystem.LayerFilesystem.mounts",
+    property(MagicMock(return_value={"/tmp/example": None, "/mnt/example": None})),
+)
```
Comment (Schamper) [inline tests/plugins/os/unix/test_trash.py:85]:
```suggestion
```

File: tests/plugins/os/unix/test_trash.py [hunk 7]
```diff
@@ -76,3 +77,29 @@ def test_gnome_trash(target_unix_users: Target, fs_unix: VirtualFilesystem) -> N
...
+    plugin = GnomeTrashPlugin(target_unix_users)
+
+    assert sorted([str(t) for _, t in plugin.trashes]) == [
+        "/home/user/.local/share/Trash",
+        "/media/user/example/.Trash-1000",
+        "/mnt/example/.Trash-1234",
+        "/tmp/example/.Trash-5678",
+    ]
+
+    assert len(list(plugin.trash())) == 11 * len(plugin.trashes)
```
Comment (Schamper) [inline tests/plugins/os/unix/test_trash.py:105]:
```suggestion
    with patch.object(target_unix_users.fs, "mounts", {"/tmp/example": None, "/mnt/example": None}):
        target_unix_users.add_plugin(UnixPlugin)
        plugin = GnomeTrashPlugin(target_unix_users)
    
        assert sorted([str(t) for _, t in plugin.trashes]) == [
            "/home/user/.local/share/Trash",
            "/media/user/example/.Trash-1000",
            "/mnt/example/.Trash-1234",
            "/tmp/example/.Trash-5678",
        ]
    
        assert len(list(plugin.trash())) == 11 * len(plugin.trashes)
```

File: dissect/target/filesystem.py [hunk 8]
```diff
@@ -1300,6 +1300,8 @@ def symlink(self, src: str, dst: str) -> None:
...
+
```
Comment (Schamper) [inline dissect/target/filesystem.py:1304]:
```suggestion
```

File: dissect/target/filesystem.py (hunk 1 above)
Comment (Schamper) [reply dissect/target/filesystem.py:1303]: My last suggestion seems to work?

**Review by Schamper — CHANGES_REQUESTED** [review]

**Review by Horofic — DISMISSED** [review]

**Review by Horofic — APPROVED** [review]

File: dissect/target/plugins/os/unix/trash.py [hunk 9]
```diff
@@ -31,15 +31,25 @@ class GnomeTrashPlugin(Plugin):
...
             for trash_path in self.PATHS:
                 if (path := user_details.home_path.joinpath(trash_path)).exists():
                     yield user_details, path
 
+        # mounted devices trash folders
+        for mount_path in list(self.target.fs.mounts.keys()) + ["/mnt", "/media"]:
+            if mount_path == "/":
+                continue
+
+            for mount_trash in self.target.fs.path(mount_path).glob("**/.Trash-*"):
```
[the remark this answers is not in the document]
Comment (Schamper) [reply dissect/target/plugins/os/unix/trash.py:50]: Are you sure? Doesn't this have the same behaviour? 

https://docs.python.org/3/library/pathlib.html#pathlib.Path.rglob

The documentation for `rglob` explicitly states it's basically the same as `.glob("**/<pattern>")`. Don't you mean `*/.Trash-*`?

**Review by Schamper — CHANGES_REQUESTED** [review]

**Review by Schamper — APPROVED** [review]

**Pull request merged**